Unencrypted support connections plus proxy support

Thaodan opened the pull request:
+ Contributes to #30.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added support for both secure (TLS) and non-secure (plain TCP) connections when connecting to Rocket.Chat.
  * Introduced new account options for selecting connection security and specifying a proxy.
  * Improved proxy support and flexible server configuration.

* **Improvements**
  * Enhanced memory management for connection-related settings.
  * Updated plugin summary and description for better clarity.

* **Bug Fixes**
  * Minor compatibility fixes for improved stability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Thaodan commented:
TLS connection don't work with right because SSL-Socket can't established:

``` irc-log
(08:55:37) GLib-GObject: ../glib/gobject/gsignal.c:2613: signal 'destroy' is invalid for instance '0x555555d955c0' of type 'GtkStatusIcon'
(08:55:37) gtkblist: added visibility manager: 1
(08:55:37) docklet: GTK+ created
(08:55:37) gtkmedia: Registering media element types
(08:55:37) util: Reading file blist.xml from directory /home/thaodan/.purple
(08:55:37) plugins: probing /usr/local/lib/purple-2/ssl-gnutls.so
(08:55:37) plugins: Unable to find saved plugin /usr/local/lib/purple-2/ssl-gnutls.so
(08:55:37) plugins: Loading saved plugin /usr/lib/purple-2/ssl-nss.so
(08:55:37) plugins: Loading saved plugin /usr/lib/purple-2/ssl.so
(08:55:37) pounce: Error reading pounces: Datei »/home/thaodan/.purple/pounces.xml« konnte nicht geöffnet werden: Datei oder Verzeichnis nicht gefunden
(08:55:37) Session Management: ICE initialized.
(08:55:37) Session Management: Connecting with no previous ID
(08:55:37) Session Management: Handling new ICE connection...
(08:55:37) done.
(08:55:37) Session Management: Connected to manager (KDE) with client ID 106f64696e000166408533700007218790093
(08:55:37) Session Management: Using /usr/bin/pidgin as command
[New Thread 0x7fffd56846c0 (LWP 916898)]
[Thread 0x7fffd56846c0 (LWP 916898) exited]
[New Thread 0x7fffd56846c0 (LWP 916899)]
[New Thread 0x7fffd4bae6c0 (LWP 916900)]
[Thread 0x7fffd56846c0 (LWP 916899) exited]
[New Thread 0x7fffd56846c0 (LWP 916901)]
[New Thread 0x7fffbbfab6c0 (LWP 916902)]
[Thread 0x7fffd56846c0 (LWP 916901) exited]
[Thread 0x7fffd4bae6c0 (LWP 916900) exited]
[Thread 0x7fffbbfab6c0 (LWP 916902) exited]
(08:55:37) g_log: Error loading plugin: libnuspell.so.5: Kann die Shared-Object-Datei nicht öffnen: Datei oder Verzeichnis nicht gefunden

[New Thread 0x7fffbbfab6c0 (LWP 916903)]
[New Thread 0x7fffd4bae6c0 (LWP 916904)]
[Thread 0x7fffbbfab6c0 (LWP 916903) exited]
[New Thread 0x7fffbbfab6c0 (LWP 916905)]
[New Thread 0x7fffd56846c0 (LWP 916906)]
[Thread 0x7fffd4bae6c0 (LWP 916904) exited]
[Thread 0x7fffbbfab6c0 (LWP 916905) exited]
[Thread 0x7fffd56846c0 (LWP 916906) exited]
[New Thread 0x7fffd56846c0 (LWP 916907)]
[New Thread 0x7fffbbfab6c0 (LWP 916908)]
[Thread 0x7fffd56846c0 (LWP 916907) exited]
[New Thread 0x7fffd56846c0 (LWP 916909)]
[New Thread 0x7fffd4bae6c0 (LWP 916910)]
[Thread 0x7fffbbfab6c0 (LWP 916908) exited]
[Thread 0x7fffd56846c0 (LWP 916909) exited]
[New Thread 0x7fffd56846c0 (LWP 916911)]
[New Thread 0x7fffbbfab6c0 (LWP 916912)]
[Thread 0x7fffd4bae6c0 (LWP 916910) exited]
[Thread 0x7fffd56846c0 (LWP 916911) exited]
[Thread 0x7fffbbfab6c0 (LWP 916912) exited]
(08:55:37) account: Connecting to account Thaodan|open.rocket.chat.
(08:55:37) connection: Connecting. gc = 0x55555682e190
(08:55:37) rocketchat: connection-security: tls
(08:55:37) rocketchat: Fetching url https://open.rocket.chat/api/me
(08:55:37) util: requesting to fetch a URL
(08:55:37) dnsquery: Performing DNS lookup for open.rocket.chat
(08:55:37) util: requesting to fetch a URL
(08:55:37) dnsquery: Performing DNS lookup for 192.168.188.1
(08:55:37) Session Management: Received first save_yourself
(08:55:37) dns: Debugger detected, performing useless query...
[Detaching after fork from child process 916915]
(08:55:37) dns: Created new DNS child 916915, there are now 1 children.
(08:55:37) dns: Successfully sent DNS request to child 916915
(08:55:37) mediamanager: Registered Audio/Source device HyperX SoloCast Analog Stereo
(08:55:37) Session Management: Received save_complete
(08:55:37) dnsquery: IP resolved for 192.168.188.1
(08:55:37) proxy: Attempting connection to 192.168.188.1
(08:55:37) proxy: Connecting to 192.168.188.1:49000 with no proxy
(08:55:37) proxy: Connection in progress
(08:55:37) g_log: Name com.canonical.AppMenu.Registrar on the session bus is owned by :1.15

(08:55:37) mediamanager: Registered Audio/Source device Webcam C270 Mono
(08:55:37) mediamanager: Registered Audio/Sink device Navi 21/23 HDMI/DP Audio Controller Digital Stereo (HDMI 4)
(08:55:37) mediamanager: Registered Audio/Sink device USB Audio Analog Surround 5.1
(08:55:37) mediamanager: Registered Audio/Sink device USB Audio Digital Stereo (IEC958)
(08:55:37) mediamanager: Registered Audio/Sink device Xperia 1
(08:55:37) mediamanager: Registered Video/Source device UVC Camera (046d:0825)
(08:55:37) mediamanager: Registered Video/Sink device AMD Radeon RX 6900 XT (RADV NAVI21)
(08:55:37) proxy: Connecting to 192.168.188.1:49000.
(08:55:37) proxy: Connected to 192.168.188.1:49000.
(08:55:37) util: request constructed
(08:55:37) docklet: embedded
[New Thread 0x7fffbbfab6c0 (LWP 916917)]
(08:55:38) dns: Got response for 'open.rocket.chat'
(08:55:38) dnsquery: IP resolved for open.rocket.chat
(08:55:38) proxy: Attempting connection to 104.22.23.195
(08:55:38) proxy: Connecting to open.rocket.chat:443 with no proxy
(08:55:38) proxy: Connection in progress
(08:55:38) util: Response headers: 'HTTP/1.1 200 OK
Cache-Control: max-age=120
Connection: close
Content-Length: 3413
Content-Type: text/xml
Date: Sun, 25 Sep 2022 05:55:38 GMT
ETag: "029A556867954EC57"
Expires: Sun, 25 Sep 2022 05:57:38 GMT
Last-Modified: Thu, 01 Jan 1970 00:01:03 GMT
Mime-Version: 1.0

'
(08:55:38) util: parsed 3413
(08:55:38) util: requesting to fetch a URL
(08:55:38) dnsquery: Performing DNS lookup for 192.168.188.1
(08:55:38) dnsquery: Performing DNS lookup for 192.168.188.1
(08:55:38) dnsquery: IP resolved for 192.168.188.1
(08:55:38) proxy: Attempting connection to 192.168.188.1
(08:55:38) proxy: Connecting to 192.168.188.1:49000 with no proxy
(08:55:38) proxy: Connection in progress
(08:55:38) dnsquery: IP resolved for 192.168.188.1
(08:55:38) proxy: Attempting connection to 192.168.188.1
(08:55:38) proxy: Connecting to 192.168.188.1:49000 with no proxy
(08:55:38) proxy: Connection in progress
(08:55:38) proxy: Connecting to 192.168.188.1:49000.
(08:55:38) proxy: Connected to 192.168.188.1:49000.
(08:55:38) util: request constructed
(08:55:38) proxy: Connecting to 192.168.188.1:49000.
(08:55:38) proxy: Connected to 192.168.188.1:49000.
(08:55:38) upnp: Local IP: 192.168.188.119
(08:55:38) proxy: Connecting to open.rocket.chat:443.
(08:55:38) proxy: Connected to open.rocket.chat:443.
(08:55:38) util: Response headers: 'HTTP/1.1 200 OK
DATE: Sun, 25 Sep 2022 05:55:38 GMT
SERVER: FRITZ!Box 7590 UPnP/1.0 AVM FRITZ!Box 7590 154.07.29
CONNECTION: keep-alive
CONTENT-LENGTH: 379
CONTENT-TYPE: text/xml; charset="utf-8"
EXT:

'
(08:55:38) util: parsed 379
(08:55:38) upnp: NAT Returned IP: 176.72.133.76
(08:55:38) nss: SSL version 3.4 using 128-bit AES-GCM with 128-bit AEAD MAC
Server Auth: 256-bit TLS 1.3, Key Exchange: 255-bit TLS 1.3, Compression: NULL
Cipher Suite Name: TLS_AES_128_GCM_SHA256
(08:55:38) nss: subject=CN=sni.cloudflaressl.com,O="Cloudflare, Inc.",L=San Francisco,ST=California,C=US issuer=CN=Cloudflare Inc ECC CA-3,O="Cloudflare, Inc.",C=US
(08:55:38) nss: subject=CN=Cloudflare Inc ECC CA-3,O="Cloudflare, Inc.",C=US issuer=CN=Baltimore CyberTrust Root,OU=CyberTrust,O=Baltimore,C=IE
(08:55:38) nss: partial certificate chain
(08:55:38) certificate/x509/tls_cached: Starting verify for open.rocket.chat
(08:55:38) certificate/x509/tls_cached: Checking for cached cert...
(08:55:38) certificate/x509/tls_cached: ...Found cached cert
(08:55:38) nss/x509: Loading certificate from /home/thaodan/.purple/certificates/x509/tls_peers/open.rocket.chat
(08:55:38) certificate/x509/tls_cached: Peer cert matched cached
(08:55:38) nss/x509: Exporting certificate to /home/thaodan/.purple/certificates/x509/tls_peers/open.rocket.chat
(08:55:38) util: Writing file /home/thaodan/.purple/certificates/x509/tls_peers/open.rocket.chat
(08:55:38) nss: Trusting CN=sni.cloudflaressl.com,O="Cloudflare, Inc.",L=San Francisco,ST=California,C=US
(08:55:38) certificate: Successfully verified certificate for open.rocket.chat
(08:55:38) util: request constructed
(08:55:38) util: Response headers: 'HTTP/1.1 200 OK
Date: Sun, 25 Sep 2022 05:55:38 GMT
Content-Type: text/html; charset=utf-8
Connection: close
content-security-policy: default-src 'self' https://open.rocket.chat; connect-src *; font-src 'self' https://open.rocket.chat data:; frame-src *; img-src * data: blob:; media-src * data:; script-src 'self' 'unsafe-eval' 'sha256-jqxtvDkBbRAl9Hpqv68WdNOieepg8tJSYu1xIy7zT34=' https://open.rocket.chat https://appleid.cdn-apple.com; style-src 'self' 'unsafe-inline' https://open.rocket.chat
vary: Accept-Encoding
vary: Accept-Encoding
x-content-type-options: nosniff
x-instance-id: pZdtYkpjeg4QFzJZK
x-xss-protection: 1
CF-Cache-Status: DYNAMIC
Server: cloudflare
CF-RAY: 750183130b4e376e-HEL
alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400

'
(08:55:38) dnsquery: Performing DNS lookup for open.rocket.chat
(08:55:38) signals: Signal data for chat-conversation-typing not found!
(08:55:38) dns: Successfully sent DNS request to child 916915
(08:55:38) dns: Got response for 'open.rocket.chat'
(08:55:38) dnsquery: IP resolved for open.rocket.chat
(08:55:38) proxy: Attempting connection to 104.22.23.195
(08:55:38) proxy: Connecting to open.rocket.chat:443 with no proxy
(08:55:38) proxy: Connection in progress
(08:55:38) proxy: Connecting to open.rocket.chat:443.
(08:55:38) proxy: Connected to open.rocket.chat:443.
(08:55:38) nss: SSL version 3.4 using 128-bit AES-GCM with 128-bit AEAD MAC
Server Auth: 256-bit TLS 1.3, Key Exchange: 255-bit TLS 1.3, Compression: NULL
Cipher Suite Name: TLS_AES_128_GCM_SHA256
(08:55:38) nss: subject=CN=sni.cloudflaressl.com,O="Cloudflare, Inc.",L=San Francisco,ST=California,C=US issuer=CN=Cloudflare Inc ECC CA-3,O="Cloudflare, Inc.",C=US
(08:55:38) nss: partial certificate chain
(08:55:38) certificate/x509/tls_cached: Starting verify for open.rocket.chat
(08:55:38) certificate/x509/tls_cached: Checking for cached cert...
(08:55:38) certificate/x509/tls_cached: ...Found cached cert
(08:55:38) nss/x509: Loading certificate from /home/thaodan/.purple/certificates/x509/tls_peers/open.rocket.chat
(08:55:38) certificate/x509/tls_cached: Peer cert matched cached
(08:55:38) nss/x509: Exporting certificate to /home/thaodan/.purple/certificates/x509/tls_peers/open.rocket.chat
(08:55:38) util: Writing file /home/thaodan/.purple/certificates/x509/tls_peers/open.rocket.chat
(08:55:38) nss: Trusting CN=sni.cloudflaressl.com,O="Cloudflare, Inc.",L=San Francisco,ST=California,C=US
(08:55:38) certificate: Successfully verified certificate for open.rocket.chat
(08:55:38) g_log: file librocketchat.c: line 2695 (rc_socket_connected_ssl): should not be reached
(08:55:42) util: Writing file prefs.xml to directory /home/thaodan/.purple
(08:55:42) util: Writing file /home/thaodan/.purple/prefs.xml
(08:55:43) util: Writing file accounts.xml to directory /home/thaodan/.purple
(08:55:43) util: Writing file /home/thaodan/.purple/accounts.xml
(08:55:43) util: Writing file blist.xml to directory /home/thaodan/.purple
(08:55:43) util: Writing file /home/thaodan/.purple/blist.xml
[Thread 0x7ffff10d46c0 (LWP 916891) exited]
dns[916915]: nobody needs me... =(
[New Thread 0x7ffff10d46c0 (LWP 917104)]
[New Thread 0x7fffd56846c0 (LWP 917105)]
[Thread 0x7ffff10d46c0 (LWP 917104) exited]
[Thread 0x7fffd56846c0 (LWP 917105) exited]
[New Thread 0x7fffd56846c0 (LWP 917108)]
[New Thread 0x7ffff10d46c0 (LWP 917109)]
[Thread 0x7fffd56846c0 (LWP 917108) exited]
[Thread 0x7ffff10d46c0 (LWP 917109) exited]
```

Thaodan committed:
Add option to specify proxy connection

The option allows the use of a proxy server that is used instead of
the real server. Since the real server is still needed inside `rc_socket_upgrade()`
we still have to keep it for that.

Signed-off-by: Björn Bidar <bjorn.bidar@thaodan.de>

Thaodan commented:
Everything works now except that in Bitlbee the select encryption option somehow doesn't work, have to check.

Thaodan commented:
Works now in Bitlbee too, Bitlbee didn't like the reverse option [list](https://github.com/EionRobb/purple-rocketchat/pull/32/commits/248c8a4f256791bc73de021cc8d91c2c346a263c#diff-af8277caeac252980019401574975ddbea7ca0ca7cd19e2f86dcf8b28855ecc2R3784).

EionRobb commented:
Might also want to check that all the tabs didn't turn into spaces before you mark it as ready to merge :)

EionRobb commented:
> Are you open to adding editorconfig to the repo to avoid such issues in the future?

Sounds good :)

Thaodan committed:
Add editorconfig to automatically set text-editor settings

Read below for more:
https://editorconfig.org/

Signed-off-by: Björn Bidar <bjorn.bidar@thaodan.de>

Thaodan committed:
Use g_memdump2 for glib 2.68 or greater

g_memdump is deprecated in glib 2.68.0.

Signed-off-by: Björn Bidar <bjorn.bidar@thaodan.de>

Thaodan committed:
Set PluginInfo summary and description for purple2 too

Signed-off-by: Björn Bidar <bjorn.bidar@thaodan.de>

EionRobb commented on librocketchat.c:
```diff
@@ -23,6 +23,10 @@
 #endif /* 2.32.0 */
 
+#if !GLIB_CHECK_VERSION(2, 68, 0)
+#define g_memdump2 g_memdup
```
Should this be `gmemdup2` instead of "dump"?

EionRobb replied:
Oh, also, needs a ` && !PURPLE_VERSION_CHECK(2, 14, 2)` because it accidentally(?) ended up in purple headers in 2.14.2

Thaodan replied:
Oh sorry typo.

What do you mean? I don't find the function in 2.14.2 sources.

EionRobb replied:
2.14.2 added a new libpurple/glibcompat.h which included a `static inline gpointer g_memdup2`

EionRobb commented on librocketchat.c:
```diff
@@ -805,6 +837,11 @@ rc_fetch_url(RocketChatAccount *ya, const gchar *url, const gchar *postdata, Roc
 #if PURPLE_VERSION_CHECK(3, 0, 0)
 	
 	PurpleHttpRequest *request = purple_http_request_new(url);
+    if (ya->websocket_server != ya->server) {
```
Does this need to be a string comparasion instead, eg `!purple_strequal(ya->websocket_server, ya->server)`?

Thaodan replied:
On my tests it did work without but might be better.

EionRobb commented on librocketchat.c:
```diff
@@ -837,9 +874,15 @@ rc_fetch_url(RocketChatAccount *ya, const gchar *url, const gchar *postdata, Roc
 	headers = g_string_new(NULL);
-	
+
+	if (ya->websocket_server != ya->server) {
```
Ditto string compare as above?

EionRobb commented on librocketchat.c:
```diff
@@ -837,9 +874,15 @@ rc_fetch_url(RocketChatAccount *ya, const gchar *url, const gchar *postdata, Roc
-	
+
+	if (ya->websocket_server != ya->server) {
+		host = g_strdup(ya->websocket_server);
```
Does this leak the memory of the old `host`?  Need `g_free(host)` before assign?

Thaodan replied:
Do I need to do so? host is just a plain char array without malloc created in the purple function.

EionRobb commented on librocketchat.c:
```diff
@@ -2203,7 +2271,13 @@ rc_close(PurpleConnection *pc)
+	if (ya->websocket != NULL)
+		purple_ssl_close(ya->websocket);
+	else if (ya->fd > 0) {
+		if (ya->pc->inpa)
```
is `pc->inpa` exposed through a function rather than direct struct access?  If so can we use that func for forwards compatibility?

Thaodan replied:
Purples Jabber plugin also uses the same way to access pc. The pc is part of the account structure, I'm not sure how that would change with Purple 3.x but I don't see any of that so far.

EionRobb replied:
Purple 3 makes all the eg `PurpleConnection` and `PurpleAccount` structs private

Thaodan replied:
Oh it seams I should have used `ya->inpa` instead.
Purple 3.x would do `g_source_remove(ya->inpa).`

EionRobb commented on librocketchat.c:
```diff
@@ -3636,6 +3802,27 @@ rc_add_account_options(GList *account_options)
+	ADD_VALUE(encryption_values, _("No encryption"), "no_tls");
+	ADD_VALUE(encryption_values, _("Require encryption"), "tls");
+
+#undef ADD_VALUE
```
Out of interest, was there a reason you went with a string in a dropdown vs a boolean? :)

Thaodan replied:
Mostly to keep it in line with other such menus, on the bitlbee side a drop down is also easier to see.

EionRobb commented on librocketchat.c:
```diff
@@ -3636,6 +3802,27 @@ rc_add_account_options(GList *account_options)
+                                            "connection_security", encryption_values);
+	account_options = g_list_append(account_options, option);
+
+	option = purple_account_option_string_new(N_("Proxy"), "proxy", "");
```
There's proxy options available in the libpurple api that can be used instead from the proxy tab using the `PurpleProxyInfo` stuff in `purple_proxy_get_setup()` https://docs.imfreedom.org/pidgin2/proxy_8h.html#a6995e6f63718dcd8644b9d4ef471392a  if that makes things easier?

Thaodan replied:
The thing is that isn't really a "proxy", it's more like using another host instead of the real host.
It works by using either stunnel or socat as an ssl proxy - both work fine from my tests.

EionRobb replied:
Oh right!  This is so you can add in the client cert?  Just putting 2 and 2 together now.  With a label like "proxy" I wouldn't want anyone to get tripped up with the other Proxy tab. 

Naming stuff is hard.

What about "intermediate server"?

Thaodan replied:
> What about "intermediate server"?

Isn't that also a proxy? I think proxy is accurate.

Thaodan committed:
Refactor and handle more errno's in rc_socket_got_data

- Always initialize variables, ping_frame_len was uninitialized in
  some cases.
- Remove redundant else after if websocket closed switch to goto
  try_reconnect section.
- Handle also errno EINPROGRESS and ENOENT
- Indent

Signed-off-by: Björn Bidar <bjorn.bidar@thaodan.de>

Thaodan committed:
Support unencrypted connections e.g. for use with a proxy

- Abstract all operations that affect sockets into small wrappers
- Initialize either TLS or proxy connection depending on the setting

Signed-off-by: Björn Bidar <bjorn.bidar@thaodan.de>

EionRobb committed:
Merge branch 'master' into thaodan/connection_improvements

EionRobb committed:
Response to coderabbit suggestions
